File: ios/Synapse/Core/Config/AppConfig.swift
import Foundation

/// Where this build points, read once from the bundle.
///
/// The three values arrive from `Config/Base.xcconfig` and the gitignored
/// `Config/Secrets.xcconfig` by way of `Info.plist`. They are stored as bare
/// hosts because an xcconfig treats `//` as the start of a comment, which would
/// silently truncate `https://example.com` to `https:`. The scheme is added
/// here instead, in one place.
enum AppConfig {

    /// What is wrong with the configuration, phrased for the person who can fix it.
    enum Problem: Equatable {
        case missingSupabaseHost
        case missingSupabaseKey
        case placeholderValues

        var message: String {
            switch self {
            case .missingSupabaseHost, .missingSupabaseKey:
                "Supabase isn't configured for this build."
            case .placeholderValues:
                "This build still has the example Supabase values."
            }
        }

        var fix: String {
            "Copy ios/Config/Secrets.example.xcconfig to ios/Config/Secrets.xcconfig "
                + "and fill in SUPABASE_HOST and SUPABASE_ANON_KEY from the Supabase "
                + "dashboard (Project Settings → API)."
        }
    }

    /// The Synapse API, e.g. `https://synapse.doitrous.com/api`.
    static let apiBaseURL: URL = {
        let host = string(for: "SynapseAPIHost") ?? "synapse.doitrous.com"
        // Local development runs plain HTTP on a port; anything else is a real
        // host and must be TLS.
        let scheme = host.hasPrefix("localhost") || host.hasPrefix("127.0.0.1") ? "http" : "https"
        guard let url = URL(string: "\(scheme)://\(host)/api") else {
            preconditionFailure("SynapseAPIHost is not a usable host: \(host)")
        }
        return url
    }()

    static var supabaseURL: URL? {
        guard let host = string(for: "SupabaseHost"), !isPlaceholder(host) else { return nil }
        return URL(string: "https://\(host)")
    }

    static var supabaseAnonKey: String? {
        guard let key = string(for: "SupabaseAnonKey"), !isPlaceholder(key) else { return nil }
        return key
    }

    /// `nil` when the build is usable. Checked at launch so a misconfigured
    /// build explains itself rather than failing at the first network call.
    static var problem: Problem? {
        guard let host = string(for: "SupabaseHost") else { return .missingSupabaseHost }
        guard let key = string(for: "SupabaseAnonKey") else { return .missingSupabaseKey }
        if isPlaceholder(host) || isPlaceholder(key) { return .placeholderValues }
        return nil
    }

    // MARK: - Reading the bundle

    /// An unsubstituted build setting comes through as the literal `$(NAME)`,
    /// and an unset one as an empty string. Both mean "not configured".
    private static func string(for key: String) -> String? {
        guard let raw = Bundle.main.object(forInfoDictionaryKey: key) as? String else { return nil }
        let value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty, !value.hasPrefix("$(") else { return nil }
        return value
    }

    private static func isPlaceholder(_ value: String) -> Bool {
        value.hasPrefix("your-")
    }
}

File: android/app/src/main/java/com/synapse/android/core/config/AppConfig.kt
package com.synapse.android.core.config

import com.synapse.android.BuildConfig

/**
 * What the build was configured with, and what it is missing.
 *
 * Missing configuration is a screen, not a crash. A fresh clone has no
 * `secrets.properties`, and an app that refused to launch there would be
 * reported as broken rather than as unconfigured.
 */
class AppConfig(
    private val rawSupabaseHost: String,
    val supabaseAnonKey: String,
    rawApiBaseUrl: String,
) {
    /** The API base with any trailing slash removed, so path joins never double up. */
    val apiBaseUrl: String = rawApiBaseUrl.trim().trimEnd('/')

    /**
     * The full Supabase URL.
     *
     * The host is stored without a scheme because that is how the iOS
     * xcconfig has to store it — an xcconfig reads `//` as a comment and
     * silently truncates a pasted URL to `https:`. Both clients therefore
     * keep the host bare and add the scheme back here.
     */
    val supabaseUrl: String
        get() {
            val host = rawSupabaseHost.trim().trimEnd('/')
            return if (host.startsWith("http://") || host.startsWith("https://")) host else "https://$host"
        }

    val missing: List<String>
        get() = buildList {
            if (rawSupabaseHost.isBlank()) add("SUPABASE_HOST")
            if (supabaseAnonKey.isBlank()) add("SUPABASE_ANON_KEY")
            if (apiBaseUrl.isBlank()) add("API_BASE_URL")
        }

    val isConfigured: Boolean get() = missing.isEmpty()

    companion object {
        fun fromBuild(): AppConfig = AppConfig(
            rawSupabaseHost = BuildConfig.SUPABASE_HOST,
            supabaseAnonKey = BuildConfig.SUPABASE_ANON_KEY,
            rawApiBaseUrl = BuildConfig.API_BASE_URL,
        )
    }
}
